surface/model_openapiv3.go

```diff
@@ -185,40 +233,75 @@ func (b *OpenAPI3Builder) buildTypeFromParameters(
 			}
 			f.Serialize = true
 			t.addField(&f)
+		} else if parameterRef := parametersItem.GetReference(); parameterRef != nil {
+			f.Type = typeForRef(parameterRef.GetXRef())
+			f.Name = strings.ToLower(f.Type)
+			f.Serialize = false
+			f.Kind = FieldKind_REFERENCE
+			t.addField(&f)
 		}
 	}
+
+	_, err = b.buildTypeFromRequestBody(name, requestBody, t)
+
+	if len(t.Fields) > 0 {
+		b.model.addType(t)
+		return t.Name, err
+	}
+	return "", err
+}
+
+// buildTypeFromRequestBody builds a service type description from the request bodies of an OpenAPI
+// description. If tIn is not given, a new type is created. Otherwise tIn is used.
+func (b *OpenAPI3Builder) buildTypeFromRequestBody(name string, requestBody *openapiv3.RequestBodyOrReference, tIn *Type) (tOut *Type, err error) {
+	if tIn == nil {
+		tOut = &Type{}
+		tOut.Name = name
+	} else {
+		tOut = tIn
+	}
+
 	if requestBody != nil {
 		content := requestBody.GetRequestBody().GetContent()
+		var f Field
+		f.Position = Position_BODY
+		f.Serialize = true
+
 		if content != nil {
-			for _, pair2 := range content.GetAdditionalProperties() {
-				if pair2.Name != "application/json" {
-					log.Printf("unimplemented: %q requestBody(%s)", name, pair2.Name)
+			for _, pair := range content.GetAdditionalProperties() {
+				if pair.Name != "application/json" {
+					log.Printf("unimplemented: %q requestBody(%s)", name, pair.Name)
 					continue
 				}
-				var f Field
-				f.Position = Position_BODY
-				f.Kind, f.Type, f.Format = b.typeForSchemaOrReference(pair2.GetValue().GetSchema())
+				f.Kind, f.Type, f.Format = b.typeForSchemaOrReference(pair.GetValue().GetSchema())
 				f.Name = strings.ToLower(f.Type) // use the schema name as the parameter name, since none is directly specified
-				f.Serialize = true
-				t.addField(&f)
+				tOut.addField(&f)
 			}
+		} else if reference := requestBody.GetReference(); reference != nil {
+			schemaOrReference := openapiv3.SchemaOrReference{&openapiv3.SchemaOrReference_Reference{Reference: reference}}
+			f.Kind, f.Type, f.Format = b.typeForSchemaOrReference(&schemaOrReference)
+			f.Name = strings.ToLower(f.Type) // use the schema name as the parameter name, since none is directly specified
+			tOut.addField(&f)
 		}
 	}
-	if len(t.Fields) > 0 {
-		b.model.addType(t)
-		return t.Name, err
-	}
-	return "", err
+
+	return tOut, err
 }
 
 // buildTypeFromResponses builds a service type description from the responses of an API method
 func (b *OpenAPI3Builder) buildTypeFromResponses(
-	m *Method,
 	name string,
-	responses *openapiv3.Responses) (typeName string, err error) {
+	responses *openapiv3.Responses,
+	fromComponent bool) (typeName string, err error) {
 	t := &Type{}
 	t.Name = name + "Responses"
 	t.Description = t.Name + " holds responses of " + name
+
+	if fromComponent {
```

Comment:
Same comment as above regarding where to put `if fromComponent`

Reply:
Done